### PR TITLE
Coreos kube-up now with less cloud init

### LINKED ISSUE #24165: Rework the kube-manifests for kube-up for coreos on gce
Previously, the coreos kube-up keeps [a private copy of master components and addon manifests](https://github.com/kubernetes/kubernetes/tree/master/cluster/gce/coreos/kube-manifests) to enable we have a working cluster to begin with. However this is not very ideal to keep the manifests up-to-date.

As salt/kube-up is still being heavily used in testing and spinning up the cluster. We should take a time to revisit how to handle this better. Maybe we can resue some functions in the trusty's [configure-helper.sh](https://github.com/kubernetes/kubernetes/blob/master/cluster/gce/trusty/configure-helper.sh) to convert the salt files on the fly. 

Or we just make salt to run on coreos in a container.

cc @kubernetes/sig-node @euank @sjpotter

### PR DESCRIPTION
**Pull request by euank**
This update includes significant refactoring. It moves almost all of the
logic into bash scripts, modeled after the `gci` cluster scripts.

The reason to do this is:
1. Avoid duplicating the saltbase manifests by reusing gci's parsing logic (easier maintenance)
2. Take an incremental step towards sharing more code between gci/trusty/coreos, again for better maintenance
3. Pave the way for making future changes (e.g. improved rkt support, kubelet support) easier to share

The primary differences from the gci scripts are the following:
1. Use of the `/opt/kubernetes` directory over `/home/kubernetes`
2. Support for rkt as a runtime
3. No use of logrotate
4. No use of `/etc/default/`
5. No logic related to noexec mounts or gci-specific firewall-stuff

It will make sense to move 2 over to gci, as well as perhaps a few other small improvements. That will be a separate PR for ease of review.

Ref #29720, this is a part of that because it removes a copy of them.

Fixes #24165

cc @yifan-gu 

Since this logic largely duplicates logic from the gci folder, it would be nice if someone closely familiar with that gave an OK or made sure I didn't fall into any gotchas related to that, so cc @andyzheng0831

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/33965)

### CODE REVIEW

File: cluster/gce/coreos/README.md [hunk 1]
```diff
@@ -0,0 +1,8 @@
+# CoreOS image
+
+The [CoreOS operating oystem](https://coreos.com/why/) is a Linux distribution optimized for running containers securely at scale.
```
Comment (yifan-gu) [inline cluster/gce/coreos/README.md]: s/oystem/system

**Comment by yifan-gu** [comment]
@euank Looks like we don't need `cluster/gce/coreos/configure-kubelet.sh` as well.

File: cluster/gce/coreos/configure-helper.sh [hunk 2]
```diff
@@ -0,0 +1,1147 @@
...
+    cp "${KUBE_DIR}/kube-manifests/kubernetes/gci-trusty/rescheduler.manifest" \
+       /etc/kubernetes/manifests/
+  fi
+}
+
+# Setup working directory for kubelet.
+function setup-kubelet-dir {
+    echo "Making /var/lib/kubelet executable for kubelet"
+    mount --bind /var/lib/kubelet /var/lib/kubelet/
+    mount -B -o remount,exec,suid,dev /var/lib/kubelet
```
Comment (yifan-gu) [inline cluster/gce/coreos/configure-helper.sh]: Some tailing spaces

File: cluster/gce/coreos/configure.sh [hunk 3]
```diff
@@ -0,0 +1,173 @@
+#!/bin/bash
+
+# Copyright 2015 The Kubernetes Authors.
```
Comment (yifan-gu) [inline cluster/gce/coreos/configure.sh]: 2016

File: cluster/gce/coreos/configure.sh [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+
+function download-kube-env {
+  # Fetch kube-env from GCE metadata server.
+  local -r tmp_kube_env="/tmp/kube-env.yaml"
+  curl --fail --retry 5 --retry-delay 3 --silent --show-error \
+    -H "X-Google-Metadata-Request: True" \
+    -o "${tmp_kube_env}" \
+    http://metadata.google.internal/computeMetadata/v1/instance/attributes/kube-env
+  # Convert the yaml format file into a shell-style file.
+  sed 's/: /=/' < "${tmp_kube_env}" > "${KUBE_DIR}/kube-env"
```
Comment (yifan-gu) [inline cluster/gce/coreos/configure.sh]: @euank Have you verified that this will always work for today's kube env yamls?

**Comment by yifan-gu** [comment]
LGTM other than the comments above. Thanks for the work!!! @euank

File: cluster/gce/coreos/configure.sh (hunk 4 above)
Comment (euank) [inline cluster/gce/coreos/configure.sh]: It does work as best as I can tell based on sourcing the file and printenv and based on skimming it.

Any specific concerns?

**Comment by euank** [comment]
As a hint for reviewers, doing a (vim)diff between the `gci` and the `coreos` versions of `configure.sh` and `configure-helper.sh` is incredibly useful for review.

File: cluster/gce/coreos/configure.sh (hunk 4 above)
Comment (yifan-gu) [inline cluster/gce/coreos/configure.sh]: @euank Thank you.

**Comment by euank** [comment]
Addressed your comments and manually tested. Once I re-added the hostnamectl thing you had, a lot of the issues I had just went away.

I don't know of anything wrong with this now. My previous review tip of vim-diffing with gci still applies as a way to make reviewing this big diff much easier.

**Comment by euank** [comment]
@s-urbaniak I don't know if you have any familiarity with the `cluster/gce/*` stuff, but if you'd be able to review it'd be appreciated!

**Comment by euank** [comment]
cc @roberthbailey @mikedanese, since you two are listed as owners in `./cluster`. Any review you're willing/able to give is appreciated!

**Comment by yifan-gu** [comment]
LGTM for the changes under `cluster/gce/coreos` dir

**Comment by roberthbailey** [comment]
can you please squash your commits?

**Comment by euank** [comment]
@roberthbailey I've squashed them a bit more, but I think the current split is the right level of separation between the different changes needed to get this to work.

**Comment by euank** [comment]
@k8s-bot gci gce e2e test this

**Comment by yifan-gu** [comment]
@robertabbott Do you mind take another look? Feel free to assign someone else if you don't have time.

**Comment by euank** [comment]
@roberthbailey This is one of the rare cases where making smaller chunks might be harder to review actually.

Because this is largely based on the `cluster/gce/gci` code, diffing these files with those ends up being fairly readable.
I'd be happy to add a first commit prior to these that simply copies gci -> coreos and then apply these changes on top of that if you'd prefer that to diffing locally.

Unfortunately, these changes are fairly strongly intermingled

**Comment by yifan-gu** [comment]
@roberthbailey 

I think more details on the PR might be helpful:

The 1st commit is mainly copying files from `configure.sh` and `configure-helper.sh` from `cluster/gce/gci` to `cluster/gce/coreos`, with very limited changes. https://github.com/kubernetes/kubernetes/pull/33965#issuecomment-251478825

The 4th commit is removing old files.
Other commits are just several line changes.


@euank 

One suggestion to make things easier for review, would be split this PR into 2.
- PR1: Copying `configure.sh` and `configure-helper.sh`, removing all old codes.
- PR2: Modify based on the previous PR1.

**Comment by euank** [comment]
Because you can't really make a PR on another PR on github, I just went straight for PR2.

@roberthbailey To make things easier to review, I have one commit that does `rm -rf cluster/gce/coreos; cp -R cluster/gce/gci cluster/gce/coreos`.

The diff between the directories is 0 after the first commit, and the next commit layers the changes needed by CoreOS (mainly removal of things CoreOS doesn't have or care about).

You can also manually test with e.g.:
```
export KUBE_SSH_USER=core
export KUBE_OS_DISTRIBUTION=coreos
export OS_DISTRIBUTION=coreos
export KUBE_GCE_MASTER_PROJECT=coreos-cloud
export KUBE_GCE_MASTER_IMAGE=coreos-alpha-1214-0-0-v20161028
export KUBE_GCE_NODE_PROJECT=${KUBE_GCE_MASTER_PROJECT}
export KUBE_GCE_NODE_IMAGE=${KUBE_GCE_MASTER_IMAGE}

./cluster/kube-up
```

**Comment by euank** [comment]
@roberthbailey Does making the first commit a straight copy of the gci folder help for you review-wise? If not, have an idea for how this might be more reviewable?

**Comment by roberthbailey** [comment]
@euank - yes, that helps. But I'm still pretty starved for review time and I'm ooo today so this may not get reviewed before the feature freeze on monday.

**Comment by mikedanese** [comment]
@k8s-bot bazel test this, issue https://github.com/kubernetes/test-infra/issues/1152

**Comment by euank** [comment]
Rebased out the conflicts @roberthbailey ...

**Comment by mikedanese** [comment]
@k8s-bot bazel test this, issue kubernetes/test-infra#1152

**Comment by euank** [comment]
Ping @roberthbailey 

As before, it's a clean "mv" followed by needed changes in the second commit, so the second commit is the important one to review. I managed to reduce it to a diff of "+287 −511" (mainly by not renaming KUBE_HOME -> KUBE_DIR even though the latter name fits a little better).

**Comment by roberthbailey** [comment]
@euank -- sorry, i've been spending all of my review bandwidth working on 1.5 blockers. I'll circle back to this PR once those are finished (hopefully at the end of next week if we ship on time). I haven't forgotten about it.

**Comment by euank** [comment]
If you have someone else you can suggest to review this, please do!

**Comment by euank** [comment]
Ping @roberthbailey; as before, if you don't have spare cycles but can suggest someone else, that would be helpful as well.

**Comment by euank** [comment]
Ping @roberthbailey

File: build/common.sh [hunk 5]
```diff
@@ -649,6 +649,50 @@ function kube::build::start_rsyncd_container() {
   kube::log::error "Could not connect to rsync container. See build/README.md for setting up remote Docker engine."
   return 1
 }
+# This will pack kube-system manifests files for distros without using salt
+# such as GCI and Ubuntu Trusty. We directly copy manifests from
+# cluster/addons and cluster/saltbase/salt. The script of cluster initialization
+# will remove the salt configuration and evaluate the variables in the manifests.
+function kube::release::package_kube_manifests_tarball() {
```
Comment (roberthbailey) [inline build/common.sh]: Isn't this redundant with the function of the same name in build/lib/release.sh?

File: cluster/gce/coreos/configure.sh [hunk 6]
```diff
@@ -14,30 +14,10 @@
...
-# Due to the GCE custom metadata size limit, we split the entire script into two
-# files configure.sh and configure-helper.sh. The functionality of downloading
-# kubernetes configuration, manifests, docker images, and binary files are
-# put in configure.sh, which is uploaded via GCE custom metadata.
-
 set -o errexit
 set -o nounset
 set -o pipefail
 
-function set-broken-motd {
```
Comment (roberthbailey) [inline cluster/gce/coreos/configure.sh]: FYI We've found that setting the motd is really useful for debugging node startup failure.

**Comment by roberthbailey** [comment]
@euank 

Sorry for the long delay on the code review. I only two concerns:

1. It seems you've duplicated some code between shell scripts. They are likely to diverge in the future and it would be nice to consolidate them into one location.

1. Can you verify that coreos/gci are mostly copies of each other now rather than intimately sharing the node startup scripts (as appears to be the case from https://github.com/euank/kubernetes/tree/3536ecc6ff7f043d5a104b92b803881e5f7e33af/cluster/gce/coreos)? I'm concerned that we'll make a change to gci that will break the coreos startup scripts and won't notice as we don't have continuous testing of them.

**Comment by bgrant0607** [comment]
@grodrigues3 Is there a PR that adds an OWNERS file for cluster/gce/coreos? Could we break it out and get it in earlier than the rest?

**Comment by roberthbailey** [comment]
@bgrant0607 -- juju and vagrant are the only directories under cluster that have an OWNERS file.

**Comment by euank** [comment]
@roberthbailey Per the original comment, number 2. This PR is a first step towards "an incremental step towards sharing more code between gci/trusty/coreos, again for better maintenance"

I want to do that in a followup PR because that change will also be intrusive to GCI.

I don't understand what you mean by "sharing startup scripts". gci/coreos folders do not share code yet (though I want them to in the future as above).
I don't think any change in the gci folder will impact the coreos folder at this time.

**Comment by roberthbailey** [comment]
@euank - thanks. That answered by second question. My only remaining question is about the duplication of functions in shell.

**Comment by bgrant0607** [comment]
@roberthbailey I was suggesting that we create a new OWNERS file

**Comment by roberthbailey** [comment]
Yup. I asked @euank to send a quick PR to add one.

File: build/common.sh (hunk 5 above)
Comment (euank) [inline build/common.sh]: The was a rebase error. After fixing it, it's a 3-line diff in the original function

I'll update the PR after verifying things still work..

Thanks for catching it!

**Comment by euank** [comment]
@roberthbailey I addressed your comment and rebased it back on master again to pick up any missing gci changes.

I validated that it works for me with my kube-up environment of:
```
export KUBE_SSH_USER=core
export KUBE_OS_DISTRIBUTION=coreos

export KUBE_GCE_MASTER_PROJECT=coreos-cloud
export KUBE_GCE_MASTER_IMAGE=$(gcloud compute images list --project coreos-cloud --filter  coreos-alpha --format=json | jq '.[0].name' -r)
export KUBE_GCE_NODE_PROJECT=coreos-cloud
export KUBE_GCE_NODE_IMAGE=${KUBE_GCE_MASTER_IMAGE}
```

I'll send an owners pr pronto as well including myself, @yifan-gu, and @ethernetdan as owners.

**Comment by roberthbailey** [comment]
@euank looks like it needs another rebase.

**Comment by euank** [comment]
Rebased ..

**Pull request merged**